src/coreclr/jit/codegencommon.cpp

```diff
@@ -2315,7 +2315,7 @@ void CodeGen::genReportEH()
 
         CORINFO_EH_CLAUSE_FLAGS flags = ToCORINFO_EH_CLAUSE_FLAGS(HBtab->ebdHandlerType);
 
-        if (isNativeAOT && (XTnum > 0))
+        if (XTnum > 0)
         {
             // For NativeAOT, CORINFO_EH_CLAUSE_SAMETRY flag means that the current clause covers same
             // try block as the previous one. The runtime cannot reliably infer this information from
```

Comment:
I assume that we are inserting the no-ops mentioned in alternative solution for CoreCLR today, and we may be able to delete that once/if we deprecate and delete the old EH impl.

Reply:
I think it is fine to do it here.

Reply:
I am not sure I understand the comment. What would we delete after we deprecate the old EH impl? This flag is needed for the new EH.

Reply:
How is CoreCLR solving the problem addressed by the new flag today? Does JIT insert the nops as necessary to disambiguate the try blocks?

Reply:
No, JIT doesn't insert any nops for CoreCLR. It doesn't explicitly skip funclets with the same try blocks while searching for the funclet covering a specific code address. I don't know how it disambiguates them. When I was debugging the test failures in the new EH caused by the same try ranges being skipped in the EH loop, I have found that nativeAOT uses that flag and that it emits some extra marker clauses to the EH info for try ranges that are the same, but not belonging to the same try. Since I didn't want to add emitting those for coreclr, I've just used that flag directly in the EH loop.

Reply:
> JIT doesn't insert any nops for CoreCLR

Consider this example: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunHiUGAWQAUASg5ceUjFACekqd07VFimfJWqeyrbuLkAnMPKiA3Aq0BfC6rDYMYABYNhAQSgBzAK74YAOwwAUQQwGAAHDABLCD8GGFEbRR1dVX0jUjNEqWtNXTsHZ2Fg0Ijo2Pis7UqtNOE6TNyrSpy8+ycXYvComLiExqlklKZDYRQG3RaeHMsgA==

You will see two nops right after the call to `Console.WriteLine(1);`. I believe that these nops are there to disambiguate the try blocks, ie they are the alternative solution for the issue addressed by SAMETRY bit. Do you agree?

Reply:
It does not look to me that the `nop`s are the solution, since they're [not created with optimizations off](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzMMA2gFkYGABYQAJgElxkgBQ37T12MkB5MT4ILlwWADkIQL58XgAvbGCuAEoAXVNzHSYUBit3ZOMM83MMKABPIuLC6iqq0oqa2vMTRqba5gBOd3JkjVamgF9K2rBEsDsGdwBBKABzDgEeAFEEFSDeEIYYZOGqlramzvdSXt3iof6RsYn3FbWkrZ3LvbO2o+1T5/PXi7bRjHGkzuMHWm22r32BzMRxQnzav3MFwGQA===).

I believe the CoreCLR solution is to generate "duplicate" clauses, there is a large comment about them here:

https://github.com/dotnet/runtime/blob/f777fe6e7bee80ea5ce57b5ae06a0c3060b8b1c4/src/coreclr/jit/codegencommon.cpp#L2341-L2381

Reply:
The first 'nop' is because a `call` on amd64 must return to an instruction within the same EH region as the call itself.

The second `nop` is indeed to ensure the `try` region is unique. But the actual reason is because the inner `try` has a `catch` that returns to the outer region, and so we ensure that return location remains in the EH region it needs to be in. Otherwise, the catch return could be in a different EH region. That mattered because of thread abort. It's possible we don't need that anymore. And RH / .NET Native / NativeAOT never had to deal with thread abort, so probably had to invent some other solution. Btw, the JIT code inserting that second nop is https://github.com/dotnet/runtime/blob/f777fe6e7bee80ea5ce57b5ae06a0c3060b8b1c4/src/coreclr/jit/fgopt.cpp#L2973-L2979, which has a comment about thread abort. Also, clr-abi.md has a big section on thread abort codegen, esp. w.r.t. EH (it also talks about the `nop` after `call` case)

I don't understand the usage of the SAMETRY bit. It seems to be used in some NAOT processing of the JIT-provided data into an internal form, deciding whether or not to add an additional FAULT block. Is this documented? Can we add some documentation to clr-abi.md, with examples?

Reply:
Here's more documentation for duplicated clauses:

https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/clr-abi.md#duplicated-clauses

Note that NativeAOT doesn't use duplicated clauses. I don't see any explanation as to _why_. The main reason for them is due to disjoint `try` regions due to extraction of handler funclets. If an exception happens in a handler, how does the VM know what protects that handler? Presumably in NativeAOT, the stack walk finds the associated `try`, realizes that we are processing the handler for that `try`, skips it, and looks for the next enclosing `try`. In that case, it needs the `SAMETRY` bit to know to skip it when looking for the "next enclosing `try`"?